authorization/token/service/token_service.go

```diff
@@ -483,6 +556,192 @@ func (s *tokenServiceImpl) Refresh(ctx context.Context, identity *account.Identi
+	}
+
+	var currentIdentityID uuid.UUID
+	serviceAccount := authtoken.IsSpecificServiceAccount(ctx, authtoken.OsoProxy, authtoken.Tenant, authtoken.JenkinsIdler, authtoken.JenkinsProxy)
+	if serviceAccount {
+		// Extract SA ID
+		id, err := manager.ContextIdentity(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		currentIdentityID = *id
+	} else {
+		// Extract user ID
+		currentIdentity, err := c.Services().UserService().LoadContextIdentityIfNotDeprovisioned(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		currentIdentityID = currentIdentity.ID
+	}
+
+	var appResponse app.ExternalToken
+
+	linkingProvider, err := c.Factories().LinkingProviderFactory().NewLinkingProvider(ctx, currentIdentityID,
+		rest.AbsoluteURL(req, "", nil), forResource)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	osProvider, ok := linkingProvider.(provider.OpenShiftIdentityProvider)
+	if ok && serviceAccount {
+		// This is a request from OSO proxy, tenant, Jenkins Idler, or Jenkins proxy service to obtain a cluster wide token
+		return c.retrieveClusterToken(ctx, forResource, forcePull, osProvider)
+	}
+
+	externalToken, err := c.loadToken(ctx, linkingProvider, currentIdentityID)
+	if err != nil {
+		return nil, nil, err
+	}
+	if externalToken != nil {
+		updatedToken, errorResponse, err := c.updateProfileIfEmpty(ctx, forResource, req, linkingProvider, externalToken, forcePull)
+		if err != nil {
+			return nil, errorResponse, err
+		}
+
+		appResponse = app.ExternalToken{
+			Scope:          updatedToken.Scope,
+			AccessToken:    updatedToken.Token,
+			TokenType:      "Bearer", // We aren't saving the token_type in the database
+			Username:       updatedToken.Username,
+			ProviderAPIURL: linkingProvider.URL(),
+		}
+
+		return &appResponse, nil, nil
+	}
+	providerName := linkingProvider.TypeName()
+	linkURL := rest.AbsoluteURL(req, fmt.Sprintf("%s?for=%s", client.LinkTokenPath(), forResource), nil)
+	errorResponse := fmt.Sprintf("LINK url=%s, description=\"%s token is missing. Link %s account\"", linkURL, providerName, providerName)
+	return nil, &errorResponse, errors.NewUnauthorizedError("token is missing")
+}
+
+func (c *tokenServiceImpl) DeleteExternalToken(ctx context.Context, currentIdentity uuid.UUID, authURL string, forResource string) error {
+
+	providerConfig, err := c.Factories().LinkingProviderFactory().NewLinkingProvider(ctx, currentIdentity, authURL, forResource)
+	if err != nil {
+		if val, _ := errors.IsUnauthorizedError(err); val {
+			return err
+		}
+		return errors.NewInternalError(ctx, err)
+	}
+
+	// Delete from local DB
+	err = c.ExecuteInTransaction(func() error {
+		err := c.Repositories().Identities().CheckExists(ctx, currentIdentity.String())
+		if err != nil {
+			return errors.NewUnauthorizedError(err.Error())
+		}
+		tokens, err := c.Repositories().ExternalTokens().LoadByProviderIDAndIdentityID(ctx, providerConfig.ID(), currentIdentity)
+		if err != nil {
+			return err
+		}
+		if len(tokens) > 0 {
```

Comment:
you could probably skip the `if len(tokens) > 0` check, the `range` would take care of it I believe